File: java/03_advanced/src/com/snake/utils/SettingsManager.java
package com.snake.utils;



import com.snake.model.Difficulty;
import java.util.prefs.Preferences;

public class SettingsManager {
    private Preferences prefs;
    private Difficulty difficulty;
    private String playerName;
    private boolean soundEnabled;
    private boolean musicEnabled;
    private boolean wallsEnabled;
    private int highScore;
    private int maxFood;
    private double powerUpChance;
    private double cellSize;
    
    public SettingsManager() {
        prefs = Preferences.userNodeForPackage(SettingsManager.class);
        loadSettings();
    }
    
    private void loadSettings() {
        String diffName = prefs.get("difficulty", "MEDIUM");
        difficulty = Difficulty.valueOf(diffName);
        playerName = prefs.get("player_name", "Player");
        soundEnabled = prefs.getBoolean("sound_enabled", true);
        musicEnabled = prefs.getBoolean("music_enabled", true);
        wallsEnabled = prefs.getBoolean("walls_enabled", false);
        highScore = prefs.getInt("high_score", 0);
        maxFood = prefs.getInt("max_food", 5);
        powerUpChance = prefs.getDouble("powerup_chance", 0.1);
        cellSize = prefs.getDouble("cell_size", 25.0);
    }
    
    public void saveSettings() {
        prefs.put("difficulty", difficulty.name());
        prefs.put("player_name", playerName);
        prefs.putBoolean("sound_enabled", soundEnabled);
        prefs.putBoolean("music_enabled", musicEnabled);
        prefs.putBoolean("walls_enabled", wallsEnabled);
        prefs.putInt("high_score", highScore);
        prefs.putInt("max_food", maxFood);
        prefs.putDouble("powerup_chance", powerUpChance);
        prefs.putDouble("cell_size", cellSize);
    }
    
    public Difficulty getDifficulty() {
        return difficulty;
    }
    
    public void setDifficulty(Difficulty difficulty) {
        this.difficulty = difficulty;
    }
    
    public String getPlayerName() {
        return playerName;
    }
    
    public void setPlayerName(String playerName) {
        this.playerName = playerName;
    }
    
    public boolean isSoundEnabled() {
        return soundEnabled;
    }
    
    public void setSoundEnabled(boolean soundEnabled) {
        this.soundEnabled = soundEnabled;
    }
    
    public boolean isMusicEnabled() {
        return musicEnabled;
    }
    
    public void setMusicEnabled(boolean musicEnabled) {
        this.musicEnabled = musicEnabled;
    }
    
    public boolean isWallsEnabled() {
        return difficulty.hasWalls();
    }
    
    public int getHighScore() {
        return highScore;
    }
    
    public void setHighScore(int highScore) {
        if (highScore > this.highScore) {
            this.highScore = highScore;
            prefs.putInt("high_score", highScore);
        }
    }
    
    public int getMaxFood() {
        return maxFood;
    }
    
    public double getPowerUpChance() {
        return difficulty.getPowerUpChance();
    }
    
    public double getCellSize() {
        return cellSize;
    }
}

File: cpp/03_advanced/src/SettingsManager.cpp

#include "../include/SettingsManager.h"

SettingsManager::SettingsManager(QObject *parent) 
    : QObject(parent), settings("MyCompany", "SnakeGame") {}

GameEngine::Difficulty SettingsManager::getDifficulty() const {
    return static_cast<GameEngine::Difficulty>(
        settings.value("difficulty", static_cast<int>(GameEngine::Difficulty::MEDIUM)).toInt()
    );
}

void SettingsManager::setDifficulty(GameEngine::Difficulty difficulty) {
    settings.setValue("difficulty", static_cast<int>(difficulty));
}

QString SettingsManager::getPlayerName() const {
    return settings.value("player_name", "Player").toString();
}

void SettingsManager::setPlayerName(const QString& name) {
    settings.setValue("player_name", name);
}

bool SettingsManager::isSoundEnabled() const {
    return settings.value("sound_enabled", true).toBool();
}

void SettingsManager::setSoundEnabled(bool enabled) {
    settings.setValue("sound_enabled", enabled);
}

bool SettingsManager::isMusicEnabled() const {
    return settings.value("music_enabled", true).toBool();
}

void SettingsManager::setMusicEnabled(bool enabled) {
    settings.setValue("music_enabled", enabled);
}

int SettingsManager::getSoundVolume() const {
    return settings.value("sound_volume", 50).toInt();
}

void SettingsManager::setSoundVolume(int volume) {
    settings.setValue("sound_volume", qBound(0, volume, 100));
}

QString SettingsManager::getUpKey() const {
    return settings.value("up_key", "Up").toString();
}

QString SettingsManager::getDownKey() const {
    return settings.value("down_key", "Down").toString();
}

QString SettingsManager::getLeftKey() const {
    return settings.value("left_key", "Left").toString();
}

QString SettingsManager::getRightKey() const {
    return settings.value("right_key", "Right").toString();
}

void SettingsManager::setControlKeys(const QString& up, const QString& down, 
                                      const QString& left, const QString& right) {
    settings.setValue("up_key", up);
    settings.setValue("down_key", down);
    settings.setValue("left_key", left);
    settings.setValue("right_key", right);
}

QColor SettingsManager::getSnakeColor() const {
    return settings.value("snake_color", QColor(50, 255, 50)).value<QColor>();
}

void SettingsManager::setSnakeColor(const QColor& color) {
    settings.setValue("snake_color", color);
}

QColor SettingsManager::getBackgroundColor() const {
    return settings.value("background_color", QColor(30, 30, 40)).value<QColor>();
}

void SettingsManager::setBackgroundColor(const QColor& color) {
    settings.setValue("background_color", color);
}

bool SettingsManager::isGridVisible() const {
    return settings.value("grid_visible", true).toBool();
}

void SettingsManager::setGridVisible(bool visible) {
    settings.setValue("grid_visible", visible);
}

int SettingsManager::getHighScore() const {
    return settings.value("high_score", 0).toInt();
}

void SettingsManager::setHighScore(int score) {
    if (score > getHighScore()) {
        settings.setValue("high_score", score);
    }
}

int SettingsManager::getTotalPlayTime() const {
    return settings.value("total_play_time", 0).toInt();
}

void SettingsManager::setTotalPlayTime(int seconds) {
    settings.setValue("total_play_time", seconds);
}

void SettingsManager::resetToDefaults() {
    settings.clear();
}
